### PR TITLE
Very detailed step-by-step on how to add a dataset

### PR DESCRIPTION
**Pull request by thomwolf**
Add very detailed step-by-step instructions to add a new dataset to the library.

### CODE REVIEW

File: ADD_NEW_DATASET.md [hunk 1]
```diff
@@ -0,0 +1,147 @@
...
+
+2. Create your dataset folder under `datasets/<your_dataset_name>` and create your dataset script under `datasets/<your_dataset_name>/<your_dataset_name>.py`. You can check out other dataset scripts under `datasets` for some inspiration. Note on naming: the dataset class should be camel case, while the dataset name is its snake case equivalent (ex: `class BookCorpus(datasets.GeneratorBasedBuilder)` for the dataset `book_corpus`).
+
+3. **Make sure you run all of the following commands from the root of your `datasets` git clone.** To check that your dataset works correctly and to create its `dataset_infos.json` file run the command:
+
+	```bash
+	python datasets-cli test datasets/<your-dataset-folder> --save_infos --all_configs
+	```
+
+4. If the command was succesful, you should now create some dummy data. Use the following command to get in-detail instructions on how to create the dummy data:
```
Comment (lhoestq) [inline ADD_NEW_DATASET.md:83]: Maybe mention the new tool ?

File: ADD_NEW_DATASET.md [hunk 2]
```diff
@@ -0,0 +1,147 @@
...
+- Your datascript might require a difficult dummy data structure. In this case make sure you fully understand the data folder logit created by the function `_split_generators(...)` and expected by the function `_generate_examples(...)` of your dataset script. Also take a look at `tests/README.md` which lists different possible cases of how the dummy data should be created.
+
+- If the dummy data tests still fail, open a PR in the repo anyways and make a remark in the description that you need help creating the dummy data.
+
+6. Format your code. Run black and isort so that your newly added files look nice with the following command:
+
+	```bash
+	make style
+	```
+
```
Comment (lhoestq) [inline ADD_NEW_DATASET.md:121]: add a step to create the dataset card ?

File: ADD_NEW_DATASET.md [hunk 3]
```diff
@@ -0,0 +1,147 @@
...
+1. Create a new branch to hold your development changes with the name of your dataset:
+
+	```bash
+	git checkout -b a-descriptive-name-for-my-changes
+	```
+
+	**do not** work on the `master` branch.
+
+2. Create your dataset folder under `datasets/<your_dataset_name>` and create your dataset script under `datasets/<your_dataset_name>/<your_dataset_name>.py`. You can check out other dataset scripts under `datasets` for some inspiration. Note on naming: the dataset class should be camel case, while the dataset name is its snake case equivalent (ex: `class BookCorpus(datasets.GeneratorBasedBuilder)` for the dataset `book_corpus`).
```
Comment (lhoestq) [inline ADD_NEW_DATASET.md:75]: we can add the list of datasets of reference here I think

**Review by lhoestq — COMMENTED** [review]
Nice ! we can share it soon for the sprint :)

File: ADD_NEW_DATASET.md [hunk 4]
```diff
@@ -60,33 +62,87 @@ And automatically create the following additional meta-data files (using the CLI
...
+
+Now let's get coding :-)
+
+The dataset script is the main entry point to load and process the data.
+
+There is a detailed explanation on how the library and scripts are organized [here](https://huggingface.co/docs/datasets/add_dataset.html).
+
+Note on naming: the dataset class should be camel case, while the dataset short_name is its snake case equivalent (ex: `class BookCorpus(datasets.GeneratorBasedBuilder)` for the dataset `book_corpus`).
+
+To add a nwew dataset, you can start by copying one of the datasets of reference listed below. The main criteria for choosing to start from one or the other reference dataset is the format of the data files (JSON/CSV/TSV/text) and whether you need several configurations (see above explanations on configurations). Feel free to reuse parts of the code and adapt them to your case:
```
Comment (lhoestq) [inline ADD_NEW_DATASET.md:86]:
```suggestion
To add a new dataset, you can start by copying one of the datasets of reference listed below. The main criteria for choosing to start from one or the other reference dataset is the format of the data files (JSON/CSV/TSV/text) and whether you need several configurations (see above explanations on configurations). Feel free to reuse parts of the code and adapt them to your case:

```

File: ADD_NEW_DATASET.md [hunk 5]
```diff
@@ -60,33 +62,87 @@ And automatically create the following additional meta-data files (using the CLI
...
+Now that your dataset script runs and create a dataset with the format you expected.
+
+1. **Make sure you run all of the following commands from the root of your `datasets` git clone.** To check that your dataset works correctly and to create its `dataset_infos.json` file run the command:
 
 	```bash
 	python datasets-cli test datasets/<your-dataset-folder> --save_infos --all_configs
 	```
 
-4. If the command was succesful, you should now create some dummy data. Use the following command to get in-detail instructions on how to create the dummy data:
+	This first command should create a `dataset_configs.JSON` file in your dataset folder.
```
Comment (lhoestq) [inline ADD_NEW_DATASET.md:127]:
```suggestion
	This first command should create a `datasets_infos.json` file in your dataset folder.

```

File: ADD_NEW_DATASET.md [hunk 6]
```diff
@@ -0,0 +1,217 @@
...
+
+
+You are now ready to start the process of adding the dataset. We will basically create the following files:
+
+	- a dataset script, e.g. `squad.py` which will contain the code to download and pre-process the dataset
+	- a dataset cart, `README.md` which will contain general informations and tags related to the dataset
+
+And automatically create the following additional meta-data files (using the CLI tools):
+
+	- a metadata file `dataset_configs.json` which will contain checksums and informations about the dataset to garantee that the loading went fine.
```
Comment (lhoestq) [inline ADD_NEW_DATASET.md:62]:
```suggestion
	- a metadata file `dataset_infos.json` which will contain checksums and informations about the dataset to garantee that the loading went fine.

```

**Review by lhoestq — APPROVED** [review]
Looks good !

**Comment by thomwolf** [comment]
Awesome! Thanks @lhoestq

**Pull request merged**